src/lib.rs

```diff
@@ -1,3 +1,248 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Data structure measurement.
+
+extern crate libc;
+
+use libc::{c_void, size_t};
+use std::cell::{Cell, RefCell};
+use std::collections::{HashMap, LinkedList};
+#[cfg(feature = "unstable")]
+use std::collections::hash_state;
+use std::hash::Hash;
+use std::mem::{size_of, transmute};
+use std::sync::Arc;
+use std::rc::Rc;
+
+extern {
+    // Get the size of a heap block.
+    //
+    // Ideally Rust would expose a function like this in std::rt::heap, which would avoid the
+    // jemalloc dependence.
+    //
+    // The C prototype is `je_malloc_usable_size(JEMALLOC_USABLE_SIZE_CONST void *ptr)`. On some
+    // platforms `JEMALLOC_USABLE_SIZE_CONST` is `const` and on some it is empty. But in practice
+    // this function doesn't modify the contents of the block that `ptr` points to, so we use
+    // `*const c_void` here.
+    fn je_malloc_usable_size(ptr: *const c_void) -> size_t;
```

Comment:
Everywhere a pointer is given to `heap_size_of()`, a size from `std::mem::size_of` or `Vec::capacity` (and alignment from `std::mem::align_of`) could be given instead. Would that be equivalent?